R/contribution_matrix.R

```diff
@@ -0,0 +1,588 @@
+    X <- matrix(0, nrow = length(study_treatment), ncol = n_treatments)
+    colnames(X) <- c(paste0(treatments[1], ":", treatments[-1]), "B")
+  } else{
+    stop("cov_parameters must be 'Shared', 'Exchangeable', or 'Unrelated'")
+  }
+  rownames(X) <- study_treatment
+  
+  #Local function to extract the studies from a vector of strings that are in the format "(Study)Control_treatment:Treatment"
+  GetStudies <- function(x){
+    substr(x, start = 2, stop = unlist(gregexpr(")", x)) - 1)
+  }
+  
+  #Local function to extract the second treatments from a vector of strings that are in the format "(Study)Control_treatment:Treatment2" or "Reference_treatment:Treatment2"
+  GetTreatments <- function(x){
+    substr(x, start = unlist(gregexpr(":", x)) + 1, stop = nchar(x))
+  }
+  
+  #Populate the design matrix for unrelated covariate parameters
+  if(cov_parameters %in% c("Unrelated", "Exchangeable")){
+    for(i in 1:length(data_no_control$Study)){
+      #If the study contains the reference treatment...
+      if(data_no_control$Control[i] == reference){
+        #...put 1 in the treatment column...
+        X[which(GetStudies(rownames(X)) == data_no_control$Study[i] & GetTreatments(rownames(X)) == data_no_control$Treatment[i]), which(GetTreatments(colnames(X)) == data_no_control$Treatment[i])] <- 1
+        #...and put the covariate value in the treatment's covariate column
+        X[which(GetStudies(rownames(X)) == data_no_control$Study[i] & GetTreatments(rownames(X)) == data_no_control$Treatment[i]), which(GetTreatments(colnames(X)) == paste0(data_no_control$Treatment[i], "_beta"))] <- covar_centered[data_no_control$Study[i]]
+        
+        #If the study does not contain the reference treatment...
+      } else if(data_no_control$Control[i] != reference){
+        #...put 1 in the treatment column...
+        X[which(GetStudies(rownames(X)) == data_no_control$Study[i] & GetTreatments(rownames(X)) == data_no_control$Treatment[i]), which(GetTreatments(colnames(X)) == data_no_control$Treatment[i])] <- 1
+        #...put -1 in the control treatment column...
+        X[which(GetStudies(rownames(X)) == data_no_control$Study[i] & GetTreatments(rownames(X)) == data_no_control$Treatment[i]), which(GetTreatments(colnames(X)) == data_no_control$Control[i])] <- -1
+        #...put the covariate value in the treatment's covariate column...
+        X[which(GetStudies(rownames(X)) == data_no_control$Study[i] & GetTreatments(rownames(X)) == data_no_control$Treatment[i]), which(GetTreatments(colnames(X)) == paste0(data_no_control$Treatment[i], "_beta"))] <- covar_centered[data_no_control$Study[i]]
+        #...and put minus the covariate value in the control treatment's covariate column
+        X[which(GetStudies(rownames(X)) == data_no_control$Study[i] & GetTreatments(rownames(X)) == data_no_control$Treatment[i]), which(GetTreatments(colnames(X)) == paste0(data_no_control$Control[i], "_beta"))] <- -covar_centered[data_no_control$Study[i]]
```

Comment:
These blocks look like they're doing the same calculations multiple times to get the matrix indices. Can the index calculations be brought out into variables before the `if` statements, then just used in these lines?